### PR TITLE
Release 2024.1 : StructuredDatatype

### PR DESCRIPTION
**Pull request by enikao**

### CODE REVIEW

**Comment by enikao** [comment]
TO DISCUSS: Shall we define required fields as `required`?

File: src/LionWeb-CSharp-Generator/impl/LanguageConstructorGenerator.cs [hunk 1]
```diff
@@ -70,9 +71,14 @@ private IEnumerable<StatementSyntax> EntityConstructorInitialization(LanguageEnt
             case Enumeration enumeration:
                 result.AddRange(enumeration.Literals.Select(LiteralConstructorInitialization));
                 break;
-            default:
+            case StructuredDataType structuredDataType:
+                result.AddRange(structuredDataType.Fields.Select(FieldConstructorInitialization));
+                break;
+            case PrimitiveType:
                 // fall-through
```
Comment (dslmeinte) [inline src/LionWeb-CSharp-Generator/impl/LanguageConstructorGenerator.cs:78]: Is it useful to output some kind of alert, or even a comment line, about this?

File: src/LionWeb-CSharp-Generator/impl/LanguageConstructorGenerator.cs [hunk 2]
```diff
@@ -70,9 +71,14 @@ private IEnumerable<StatementSyntax> EntityConstructorInitialization(LanguageEnt
...
                 break;
+            default:
+                throw new ArgumentException($"unsupported entity: {entity}", nameof(entity));
```
Comment (dslmeinte) [inline src/LionWeb-CSharp-Generator/impl/LanguageConstructorGenerator.cs:81]: "unsupported entity [**type**]", I'd say

File: src/LionWeb-CSharp-Generator/impl/LanguageGenerator.cs [hunk 3]
```diff
@@ -139,9 +140,14 @@ private IEnumerable<MemberDeclarationSyntax> EntityLanguageMembers(LanguageEntit
             case Enumeration enumeration:
                 result.AddRange(enumeration.Literals.SelectMany(LiteralLanguageMember));
                 break;
-            default:
+                case StructuredDataType structuredDataType:
+                    result.AddRange(structuredDataType.Fields.SelectMany(FieldLanguageMember));
+                    break;
+            case PrimitiveType:
                 // fall-through
```
Comment (dslmeinte) [inline src/LionWeb-CSharp-Generator/impl/LanguageGenerator.cs:147]: (same comments as above)

File: test/LionWeb-CSharp-Test/languages/generated/Multi.g.cs [hunk 4]
```diff
@@ -25,7 +25,7 @@ public MultiLanguage(string id) : base(id)
 	/// <inheritdoc/>
         public override IReadOnlyList<LanguageEntity> Entities => [Container];
 	/// <inheritdoc/>
-        public override IReadOnlyList<Language> DependsOn => [];
+        public override IReadOnlyList<Language> DependsOn => [Examples.Library.M2.LibraryLanguage.Instance, Examples.Library.M2.LibraryLanguage.Instance];
```
Comment (dslmeinte) [inline test/LionWeb-CSharp-Test/languages/generated/Multi.g.cs:28]: Duplicate entry?

**Review by dslmeinte — APPROVED** [review]

**Comment by dslmeinte** [comment]
> TO DISCUSS: Shall we define required fields as `required`?

Discuss on Monday, or in the form of an issue?

File: src/LionWeb-CSharp/core/BaseTypes.cs [hunk 5]
```diff
@@ -241,6 +241,65 @@ bool IWritableNode.RemoveAnnotations(IEnumerable<IWritableNode> annotations) =>
...
+    /// </summary>
+    public IEnumerable<Field> CollectAllSetFields();
+
+    /// <summary>
+    /// Gets the value of the given <paramref name="field"/> on <c>this</c>.
+    /// </summary>
+    /// <exception cref="UnsetFieldException">If <paramref name="field"/> has not been set.</exception>
+    /// <see cref="CollectAllSetFields"/>
+    public object? Get(Field field);
+}
```
Comment (enikao) [inline src/LionWeb-CSharp/core/BaseTypes.cs:265]: Base interface for all StructuredDataType instances

File: src/LionWeb-CSharp/core/BaseTypes.cs [hunk 6]
```diff
@@ -241,6 +241,65 @@ bool IWritableNode.RemoveAnnotations(IEnumerable<IWritableNode> annotations) =>
...
+            return false;
+        }
+
+        return EqualityComparer<T?>.Default.Equals(this.Value, ((NullableStructMember<T>)obj).Value);
+    }
+
+    /// <inheritdoc />
+    public override int GetHashCode() => Value?.GetHashCode() ?? 0;
+}
+
```
Comment (enikao) [inline src/LionWeb-CSharp/core/BaseTypes.cs:302]: Workaround to allow recursive SDTs

File: test/LionWeb-CSharp-Test/languages/generated/SDTLang.g.cs [hunk 7]
```diff
@@ -0,0 +1,991 @@
...
+		if (_language.C.EqualsIdentity(structuredDataType))
+			return new C((string?)fieldValues.Get(_language.C_name), (D?)fieldValues.Get(_language.C_c2d), (E?)fieldValues.Get(_language.C_c2e));
+		if (_language.D.EqualsIdentity(structuredDataType))
+			return new D((string?)fieldValues.Get(_language.D_name));
+		if (_language.E.EqualsIdentity(structuredDataType))
+			return new E((string?)fieldValues.Get(_language.E_name), (F?)fieldValues.Get(_language.E_e2f), (B?)fieldValues.Get(_language.E_e2b));
+		if (_language.F.EqualsIdentity(structuredDataType))
+			return new F((string?)fieldValues.Get(_language.F_name), (C?)fieldValues.Get(_language.F_f2c));
+		throw new UnsupportedStructuredDataTypeException(structuredDataType);
+	}
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/languages/generated/SDTLang.g.cs:265]: Example SDT factory method

File: test/LionWeb-CSharp-Test/languages/generated/SDTLang.g.cs [hunk 8]
```diff
@@ -0,0 +1,991 @@
...
+	/// <inheritdoc/>
+        public Object? Get(Field field)
+	{
+		if (SDTLangLanguage.Instance.FullyQualifiedName_name.EqualsIdentity(field))
+			return Name;
+		if (SDTLangLanguage.Instance.FullyQualifiedName_nested.EqualsIdentity(field) && _nested.Value != null)
+			return Nested;
+		throw new UnsetFieldException(field);
+	}
+}
```
Comment (enikao) [inline test/LionWeb-CSharp-Test/languages/generated/SDTLang.g.cs:689]: Example of recursive SDT

**Comment by enikao** [comment]
@JojoEffect I marked the parts relevant for the design.

File: src/LionWeb-CSharp/core/BaseTypes.cs (hunk 6 above)
Comment (JojoEffect) [reply src/LionWeb-CSharp/core/BaseTypes.cs:302]: Maybe we could have a short call about this - I'm not sure what restrictions/issues you are "working around" here

Comment (enikao) [reply src/LionWeb-CSharp/core/BaseTypes.cs:302]: It's required to enable `new FullyQualifiedName(Name="a") == new FullyQualifiedName(Name="a",Nested=null)`.

**Comment by enikao** [comment]
> I've got a couple of questions about naming, especially w.r.t. the bits that are specific to support for multiple LionWeb versions. 

Happy to discuss them, naming is obviously the hardest part.

> Specifically: what does the postfix `VersionsSpecifics` mean? Is it possible to document that?

I'll add some docs. It's externalized stuff that's specific to some LW version. E.g. `IDeserializerVersionSpecifics` is stuff externalized from deserializers, because it differs between LW versions.

**Comment by enikao** [comment]
> I'll add some docs. It's externalized stuff that's specific to some LW version. E.g. `IDeserializerVersionSpecifics` is stuff externalized from deserializers, because it differs between LW versions.

Is this understandable (on `IVersionSpecifics`)?

```
/// Externalized logic of some class, specific to one version of LionWeb standard.
```

File: src/LionWeb-CSharp-Generator/impl/LanguageConstructorGenerator.cs (hunk 1 above)
Comment (enikao) [reply src/LionWeb-CSharp-Generator/impl/LanguageConstructorGenerator.cs:86]: Why? This is normal operation: We don't need to generate anything for PrimitiveTypes.

File: src/LionWeb-CSharp/core/BaseTypes.cs (hunk 6 above)
Comment (enikao) [reply src/LionWeb-CSharp/core/BaseTypes.cs]: after our decisions in https://github.com/LionWeb-io/specification/issues/265, I removed this class -- we don't allow circles in SDTs any more.

**Pull request merged**